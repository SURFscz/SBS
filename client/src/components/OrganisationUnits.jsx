import React, {useRef, useState} from "react";
import "./OrganisationUnits.scss";
import {isEmpty, splitListSemantically, stopEvent} from "../utils/Utils";
import I18n from "../locale/I18n";
import {unitUsage} from "../api";
import ConfirmationDialog from "./confirmation-dialog/ConfirmationDialog";
import {ReactComponent as TrashIcon} from "@surfnet/sds/icons/functional-icons/bin.svg";
import ErrorIndicator from "./_redesign/ErrorIndicator";
import SpinnerField from "./_redesign/SpinnerField";

export const OrganisationUnits = ({units, setUnits, setDuplicated}) => {

    const [duplicateIndex, setDuplicateIndex] = useState(-1);
    const [deletedAll, setDeletedAll] = useState(false);
    const [references, setReferences] = useState({});
    const [loading, setLoading] = useState(false);
    const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
    const [removalIndex, setRemovalIndex] = useState(-1);

    const inputRef = useRef(null);

    const focusAfterAdd = () => {
        inputRef.current && inputRef.current.focus();
    }

    const cancelRemoval = () => {
        setReferences({});
        setConfirmationDialogOpen(false);
        setRemovalIndex(-1);
    }

    const internalOnChange = index => e => {
        const value = e.target.value;
        if (units.filter(unit => unit.name.toLowerCase() === value.toLowerCase()).length > 0) {
            setDuplicateIndex(index);
            setDuplicated(true);
        } else {
            setDuplicateIndex(-1);
            setDuplicated(false);
        }
        const newUnits = [...units];
        const unit = newUnits[index];
        unit.name = value;
        newUnits.splice(index, 1, unit);
        setUnits(newUnits);
    }

    const doRemoveUnit = index => {
        units.splice(index, 1);
        const newUnits = [...units];
        setUnits(newUnits);
        if (newUnits.length === 0) {
            setDeletedAll(true);
        }
        if (newUnits.length === new Set(newUnits.map(unit => unit.name.toLowerCase())).size) {
            setDuplicateIndex(-1);
            setDuplicated(false);
        }
    }

    const hasReferences = res => {
        return !isEmpty(res.collaborations) || !isEmpty(res.invitations)
            || !isEmpty(res.collaboration_requests) || !isEmpty(res.api_keys);
    }

    const removeAfterConfirmation = () => {
        doRemoveUnit(removalIndex);
        setConfirmationDialogOpen(false);
        setReferences({});
    }

    const removeUnit = index => e => {
        stopEvent(e);
        const unit = units[index];
        if (unit.organisation_id && unit.id) {
            setLoading(true);
            unitUsage(unit).then(res => {
                setReferences(res);
                if (hasReferences(res)) {
                    setRemovalIndex(index);
                    setConfirmationDialogOpen(true);
                } else {
                    doRemoveUnit(index);
                }
                setLoading(false);
            })
        } else {
            setReferences({});
            doRemoveUnit(index);
        }
    }

    const addUnit = e => {
        stopEvent(e);
        setUnits([...units.concat({name: ""})]);
        setTimeout(() => focusAfterAdd(), 250);
    }

    const renderConfirmation = () => {
        return (
            <div className="remove-unit-confirmation">
                <p>{I18n.t("units.used")}</p>
                <ul>
                    {
                        ["collaborations", "invitations", "collaboration_requests", "organisation_memberships", "api_keys"]
                            .filter(name => !isEmpty(references[name]))
                            .map((name, index) =>
                                <li key={index}>
                                    {`${I18n.t(`units.${name}`)} (${splitListSemantically(references[name],
                                        I18n.t("service.compliancySeparator"))})`}
                                </li>
                            )
                    }
                </ul>
            </div>
        );
    }

    if (loading) {
        return <SpinnerField/>
    }

    if (!deletedAll) {
        units = units.length > 0 ? units : [{name: ""}]
    }

    const unitName = removalIndex !== -1 ? (units[removalIndex] || {}).name : "";
    return (
        <div className="organisation-units sds--text-field ">
            <ConfirmationDialog isOpen={confirmationDialogOpen}
                                cancel={cancelRemoval}
                                confirm={() => removeAfterConfirmation()}
                                question={I18n.t("units.confirmation", {name: unitName})}
                                closeTimeoutMS={0}
                                isWarning={true}>
                {confirmationDialogOpen && renderConfirmation()}
            </ConfirmationDialog>
            <label>{I18n.t("units.label")}</label>

            {units.map((unit, index) => {
                const refProps = (index + 1) === units.length ? {ref: inputRef} : {};
                return (
                    <div className={`input-field ${index === 0 ? "first" : ""}`} key={index}>
                        <div className="inner-input-field">
                            <input type="text"
                                   value={unit.name}
                                   onChange={internalOnChange(index)}
                                   className={`sds--text-field--input`}
                                   {...refProps}
                            />
                            <div className={`input-field-link input-field-delete`}>
                                <a href={"#"} onClick={removeUnit(index)}>
                                    <TrashIcon/>
                                </a>
                            </div>
                        </div>
                        {duplicateIndex === index &&
                            <ErrorIndicator msg={I18n.t("units.duplicated", {name: unit.name})}
                                            standalone={true}/>
                        }
                    </div>)
            })
            }
            <a className={"add-unit"} href="#" onClick={addUnit}>
                {I18n.t("units.add")}
            </a>

        </div>
    );
}
