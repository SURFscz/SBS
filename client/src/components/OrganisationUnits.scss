@import "../stylesheets/vars.scss";

div.organisation-units {
    display: flex;
    position: relative;
    flex-direction: column;
    margin-top: 27px;

    a.add-unit {
        margin-top: 15px;
    }
}