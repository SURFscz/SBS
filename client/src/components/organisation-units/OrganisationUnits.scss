@import "../../stylesheets/vars";

div.organisation-units {
    display: flex;
    position: relative;
    flex-direction: column;
    margin-top: 27px;

    a.add-unit {
        margin-top: 15px;
    }
}
