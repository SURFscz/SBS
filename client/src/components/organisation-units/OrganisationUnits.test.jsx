import React from "react";
import {render} from "@testing-library/react";
import {OrganisationUnits} from "./OrganisationUnits";

const units = ["unit1", "unit2"]

describe("OrganisationUnits", () => {
    it("matches snapshot with default props", () => {
        const {asFragment} = render(
            <OrganisationUnits
                units={units}
                setUnits={() => {
                }}
                setDuplicated={() => {
                }}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot with custom title", () => {
        const {asFragment} = render(
            <OrganisationUnits
                title="Custom Title"
                units={units}
                setUnits={() => {
                }}
                setDuplicated={() => {
                }}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot with custom description", () => {
        const {asFragment} = render(
            <OrganisationUnits
                description="Custom Description"
                units={units}
                setUnits={() => {
                }}
                setDuplicated={() => {
                }}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
