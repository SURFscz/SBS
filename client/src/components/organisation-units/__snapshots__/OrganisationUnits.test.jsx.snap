// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganisationUnits matches snapshot with custom description 1`] = `
<DocumentFragment>
  <div
    class="organisation-units sds--text-field "
  >
    <label>
      Unit name
    </label>
    <div
      class="input-field first"
    >
      <div
        class="inner-input-field"
      >
        <input
          class="sds--text-field--input"
          type="text"
          value=""
        />
        <div
          class="input-field-link input-field-delete"
        >
          <a
            href="#"
          >
            <svg>
              bin.svg
            </svg>
          </a>
        </div>
      </div>
    </div>
    <div
      class="input-field "
    >
      <div
        class="inner-input-field"
      >
        <input
          class="sds--text-field--input"
          type="text"
          value=""
        />
        <div
          class="input-field-link input-field-delete"
        >
          <a
            href="#"
          >
            <svg>
              bin.svg
            </svg>
          </a>
        </div>
      </div>
    </div>
    <a
      class="add-unit"
      href="#"
    >
      + Add unit
    </a>
  </div>
</DocumentFragment>
`;

exports[`OrganisationUnits matches snapshot with custom title 1`] = `
<DocumentFragment>
  <div
    class="organisation-units sds--text-field "
  >
    <label>
      Unit name
    </label>
    <div
      class="input-field first"
    >
      <div
        class="inner-input-field"
      >
        <input
          class="sds--text-field--input"
          type="text"
          value=""
        />
        <div
          class="input-field-link input-field-delete"
        >
          <a
            href="#"
          >
            <svg>
              bin.svg
            </svg>
          </a>
        </div>
      </div>
    </div>
    <div
      class="input-field "
    >
      <div
        class="inner-input-field"
      >
        <input
          class="sds--text-field--input"
          type="text"
          value=""
        />
        <div
          class="input-field-link input-field-delete"
        >
          <a
            href="#"
          >
            <svg>
              bin.svg
            </svg>
          </a>
        </div>
      </div>
    </div>
    <a
      class="add-unit"
      href="#"
    >
      + Add unit
    </a>
  </div>
</DocumentFragment>
`;

exports[`OrganisationUnits matches snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="organisation-units sds--text-field "
  >
    <label>
      Unit name
    </label>
    <div
      class="input-field first"
    >
      <div
        class="inner-input-field"
      >
        <input
          class="sds--text-field--input"
          type="text"
          value=""
        />
        <div
          class="input-field-link input-field-delete"
        >
          <a
            href="#"
          >
            <svg>
              bin.svg
            </svg>
          </a>
        </div>
      </div>
    </div>
    <div
      class="input-field "
    >
      <div
        class="inner-input-field"
      >
        <input
          class="sds--text-field--input"
          type="text"
          value=""
        />
        <div
          class="input-field-link input-field-delete"
        >
          <a
            href="#"
          >
            <svg>
              bin.svg
            </svg>
          </a>
        </div>
      </div>
    </div>
    <a
      class="add-unit"
      href="#"
    >
      + Add unit
    </a>
  </div>
</DocumentFragment>
`;
